sprimber-engine/src/main/java/com/griddynamics/qa/sprimber/engine/executor/TestCaseActionsExecutor.java

```diff
@@ -27,62 +27,74 @@ OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
-    public ExecutionResult executeStep(TestStep testStep) {
-        ExecutionResult stepResult;
-        Object targetObject = applicationContext.getBean(testStep.getStepAction().getMethod().getDeclaringClass());
-        boolean isBeforeStepsPassed = testStep.getBeforeStepActions().stream()
-                .map(this::executeHookAction)
-                .allMatch(actionExecutionResult -> actionExecutionResult.equals(ExecutionResult.PASSED));
-        try {
-            ReflectionUtils.invokeMethod(testStep.getStepAction().getMethod(), targetObject, testStep.getStepArguments());
-            stepResult = ExecutionResult.PASSED;
-        } catch (Throwable throwable) {
-            LOGGER.error(throwable.getLocalizedMessage());
-            stepResult = errorMapper.parseThrowable(throwable);
-        }
-        boolean isAfterStepsPassed = testStep.getAfterStepActions().stream()
-                .map(this::executeHookAction)
-                .allMatch(actionExecutionResult -> actionExecutionResult.equals(ExecutionResult.PASSED));
-        return isBeforeStepsPassed && isAfterStepsPassed ?
-                stepResult.equals(ExecutionResult.PASSED) ? ExecutionResult.PASSED : stepResult : ExecutionResult.FAILED;
+    public void cleanStageResults() {
+        currentStageResults.clear();
+    }
+
+    public List<ExecutionResult> currentFailures() {
+        return currentStageResults.stream()
+                .filter(executionResult -> !executionResult.getStatus().equals(PASSED))
+                .collect(Collectors.toList());
+    }
+
+    public ExecutionResult executeStepAction(TestStep testStep) {
+        return executeMethod(testStep.getStepAction().getMethod(), testStep.getStepArguments());
     }
 
     public ExecutionResult executeHookAction(ActionDefinition actionDefinition) {
-        Object targetObject = applicationContext.getBean(actionDefinition.getMethod().getDeclaringClass());
+        return executeMethod(actionDefinition.getMethod(), new Object[0]);
+    }
+
+    public ExecutionResult executeMethod(Method method, Object[] args) {
+        Object targetObject = applicationContext.getBean(method.getDeclaringClass());
         try {
-            ReflectionUtils.invokeMethod(actionDefinition.getMethod(), targetObject);
-            return ExecutionResult.PASSED;
+            ReflectionUtils.invokeMethod(method, targetObject, args);
+            ExecutionResult result = new ExecutionResult(PASSED);
+            currentStageResults.add(result);
+            // TODO: 11/15/18 currently return back the result for compatibility with event listeners
+            return result;
         } catch (Throwable throwable) {
+            ExecutionResult result = errorMapper.parseThrowable(throwable);
+            result.conditionallyPrintStacktrace();
+            currentStageResults.add(result);
+            LOGGER.trace(result.getErrorMessage());
             LOGGER.error(throwable.getLocalizedMessage());
```

Comment:
Is that the best practice to call throwable.getLocalizedMessage() method?
As far as I know - usually this thing will return null in most cases